src/Docker.DotNet/Endpoints/StreamUtil.cs

```diff
@@ -1,97 +1,61 @@
-            {
-                await client.HandleIfErrorResponseAsync(response.StatusCode, response);
-
-                using (var stream = await response.Content.ReadAsStreamAsync())
-                {
-                    // ReadLineAsync must be cancelled by closing the whole stream.
-                    using (cancel.Register(() => stream.Dispose()))
-                    {
-                        using (var reader = new StreamReader(stream, new UTF8Encoding(false)))
-                        {
-                            string line;
-                            try
-                            {
-                                while ((line = await reader.ReadLineAsync()) != null)
-                                {
-                                    var prog = client.JsonSerializer.DeserializeObject<T>(line);
-                                    if (prog == null) continue;
-
-                                    progress.Report(prog);
-                                }
-                            }
-                            catch (ObjectDisposedException)
-                            {
-                                // The subsequent call to reader.ReadLineAsync() after cancellation
-                                // will fail because we disposed the stream. Just ignore here.
-                            }
-                        }
-                    }
-                }
-            }
-        }
-
-        private static async Task<T> WithCancellation<T>(this Task<T> task, CancellationToken cancellationToken)
-        {
-            var tcs = new TaskCompletionSource<bool>();
-            using (cancellationToken.Register(s => ((TaskCompletionSource<bool>)s).TrySetResult(true), tcs))
-            {
-                if (task != await Task.WhenAny(task, tcs.Task))
-                {
-                    throw new OperationCanceledException(cancellationToken);
-                }
-            }
-
-            return await task;
-        }
-    }
-}
+using System;
+using System.IO;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+
+namespace Docker.DotNet.Models
+{
+    internal static class StreamUtil
+    {
+        private static readonly Newtonsoft.Json.JsonSerializer _serializer = new Newtonsoft.Json.JsonSerializer();
+
+        internal static async Task MonitorStreamAsync(Task<Stream> streamTask, DockerClient client, CancellationToken cancel, IProgress<string> progress)
+        {
+            var tcs = new TaskCompletionSource<string>();
+
+            using (var stream = await streamTask)
+            using (var reader = new StreamReader(stream, new UTF8Encoding(false)))
+            using (cancel.Register(() => tcs.TrySetCanceled(default)))
+            {
+                while (!cancel.IsCancellationRequested)
+                {
+                    string line = await await Task.WhenAny(reader.ReadLineAsync(), tcs.Task);
```

Comment:
Alright one last question. When `tcs.Task` is selected the 2nd `await` will throw because there is no result. I assume this is the correct behavior and we want to throw the cancellation exception?

Reply:
When cancelling a token I expect to get an OpertionCancelledException. Otherwise it would mean task reached its end instead of being cancelled. 

I have no knowledge about architecture or system designs, i just enjoy thinks working on the way I like. Do the minimum to not do it twice.

I've included several tests on this, expecting OperationCancelledException to be thrown and failing test otherwise. 
There should be on IContainerOperationsTests

Reply:
I don't think it will throw there. The `tcs.Task` will be returned.  

In this case I wouldn't assign to the `line` variable directly. I would assign it to a Task and would check which task is that - the readline or the cancellation. 

If that is the cancellation task, then yeah, it is cancelled. If it is the deadline Task, then you `string line = await readLineTask();`.

Reply:
There should be a test behind this

Reply:
Agreed.

Reply:
I listened to your advice. It's not exactly as you said but shares the same idea.
It helped fixing some problems when reading streams from containers created without a tty
Thank you @galvesribeiro